vimscript.html.markdown

```diff
@@ -0,0 +1,572 @@
+endfunction
+
+" Dictionary functions, poor man's OOP methods (:h Dictionary-function)
+function! Mylen() dict
+    return len(self.data)     |" Implicit variable self
+endfunction
+let mydict = {'data': [0, 1, 2, 3], 'len': function("Mylen")}
+echo mydict.len()
+
+" Alternatively, more concise
+let mydict = {'data': [0, 1, 2, 3]}
+function! mydict.len()
+    return len(self.data)
+endfunction
+```
+
+Without the exclamation mark it would be an error to re-define a function, with
+the exclamation mark the new definition can replace the old one. Since Vim
+script files can be reloaded several times over the course of a session it is
+best to use the exclamation mark unless you really know what you are doing.
+
+### Calling functions
+
+```vim
+" Call a function for its return value, and possibly for its side effects
+let animals = keys({'cow': 'moo', 'dog': 'woof', 'cat': 'meow'})
+
+" Call a function for its side effects only, ignore potential result
+call sign_undefine()
+```
+
+Recall that Vim script is embedded within the ex-commands, that is why we
+cannot just call a function directly, we have to use the `:call` ex-command.
+
+### Function namespaces
+
+See `:h write-library-script`, `:h autoload`
+
+```vim
+" Must be defined in autoload/foo/bar.vim
+function! foo#bar#log(value)
+    echomsg value
+endfunction
+
+call foo#bar#log('Hello')
+```
+
+
+## Frequently used ex-commands
+
+### Sourcing runtime files
+
+See `:h 'runtimepath'`
+
+```vim
+" Source first match among runtime paths
+runtime plugin/my-plugin.vim
+```
+
+### Defining new ex-commands
+See `:h 40.2`, `:h :command`
+
+```vim
+" First argument is the name of the command, rest is the command body
+command! SwapLines normal ddp
+```
+
+The exclamation mark works the same as with `:function`. User-defined commands
+must start with a capital letter. The `:command` command can take many extra
+optional arguments, such as `-nargs`, all of them start with a dash.
+
+### Defining auto-commands
+See `:h 40.3`, `:h autocmd`
+
+```vim
+" Events, patterns, commands
+autocmd Filetype html,xml :set tabstop=2
+```
+
+Events and patterns are separated by commas with no space between. See `:h
+autocmd-events` for standard events.
+
+#### Auto groups
+
+When a file is source multiple times the auto-commands are defined anew,
+without deleting the old ones, causing auto-commands to pile up over time. Use
+auto-groups and the following ritual to guard against this.
+
+```vim
+augroup sgml-like   |" The name of the group is arbitrary
+		autocmd!        |" Deletes all auto-commands in the current group
+		autocmd Filetype html,xml :set tabstop=2
+augroup END         |" Switch back to default auto-group
+```
+
+### Executing (run-time macros of sorts)
+
+Sometimes we need to construct an ex-command where part of the command is not
+known until runtime.
+
+```vim
+let line = 3               |" Line number deternimed at runtime
+execute line . 'delete'    |" Delete a line
```

Comment:
does this delete line 3? not really clear to me what the `.` is doing and if "delete" is a keyword or something.

Reply:
@karismatic-megafauna Just an FYI, this string concatenation was mentioned [here](https://github.com/adambard/learnxinyminutes-docs/pull/3763/files#diff-eca009cb1a774d802a20c34b44b8e723R122).

Reply:
I am not sure it was mentioned at the time of me commenting, I think the author rebased as my comments are out of date and their is still one commit... :/ 

Thank you for pointing it out tho!